File: code/Python/chapter14_analysis.py
import pandas as pd
import statsmodels.formula.api as smf
from statsmodels.stats.anova import anova_lm, AnovaRM

anc = pd.read_csv("data/chapter14_ancova.csv")
anc["group"] = pd.Categorical(anc["group"],
    categories=["Control","Workshop","Blended"])
model = smf.ols(
    "post_score ~ baseline_score + group",
    data=anc
).fit(cov_type="HC3")
print(model.summary())

reduced = smf.ols(
    "post_score ~ baseline_score + group",
    data=anc
).fit()
expanded = smf.ols(
    "post_score ~ baseline_score * group",
    data=anc
).fit()
print(anova_lm(reduced, expanded))

long = pd.read_csv("data/chapter14_longitudinal.csv")
long["time"] = pd.Categorical(long["time"],
    categories=["Baseline","Post","Follow-up"], ordered=True)
long["treatment"] = pd.Categorical(long["treatment"],
    categories=["Control","Intervention"])
obs = long.dropna(subset=["outcome_score"])
count = obs.groupby("student_id").time.nunique()
balanced = obs[obs.student_id.isin(count[count==3].index)]

print(AnovaRM(
    balanced,
    depvar="outcome_score",
    subject="student_id",
    within=["time"]
).fit())

mixed = smf.mixedlm(
    "outcome_score ~ time_num * treatment + I(time_num ** 2) + prior_score",
    obs,
    groups=obs["student_id"],
    re_formula="~time_num"
).fit(reml=True, method="lbfgs")
print(mixed.summary())

File: code/R/chapter14_analysis.R
library(dplyr)
library(sandwich)
library(lmtest)
library(afex)
library(lme4)
library(lmerTest)
library(emmeans)

anc <- read.csv("data/chapter14_ancova.csv")
anc$group <- factor(anc$group,
  levels=c("Control","Workshop","Blended"))
ancova <- lm(post_score ~ baseline_score + group, data=anc)
coeftest(ancova, vcov=vcovHC(ancova, type="HC3"))
anova(ancova, lm(post_score ~ baseline_score * group, data=anc))
emmeans(ancova, ~group, cov.reduce=mean)

long <- read.csv("data/chapter14_longitudinal.csv")
long$time <- factor(long$time,
  levels=c("Baseline","Post","Follow-up"))
long$treatment <- factor(long$treatment,
  levels=c("Control","Intervention"))

complete_ids <- long |>
  group_by(student_id) |>
  summarise(ok=all(!is.na(outcome_score))) |>
  filter(ok) |>
  pull(student_id)

balanced <- long |> filter(student_id %in% complete_ids)

rm <- aov_car(
  outcome_score ~ treatment*time + Error(student_id/time),
  data=balanced,
  factorize=FALSE
)
print(rm)

mixed <- lmer(
  outcome_score ~ time_num*treatment + I(time_num^2) + prior_score +
    (1+time_num|student_id),
  data=long,
  REML=TRUE,
  na.action=na.exclude
)
summary(mixed)
emmeans(mixed, ~treatment*time_num, at=list(time_num=c(0,1,2)))